src/assets/scripts/client/aircraft/AircraftModel.js

```diff
@@ -1457,18 +1545,48 @@ export default class AircraftModel {
             //     return;
 
             case MCP_MODE.ALTITUDE.VNAV: {
-                const waypointAltitude = this.fms.currentWaypoint.altitudeRestriction;
-                const waypointHasAltitude = waypointAltitude !== -1;
+                const hardRestrictedWaypointModel = this.fms.nextHardAltitudeRestrictedWaypoint;
 
-                if (!waypointHasAltitude) {
-                    return this.mcp.altitude;
+                if (_isNil(hardRestrictedWaypointModel)) {
+                    return;
                 }
 
+                let waypointAltitude = hardRestrictedWaypointModel.altitudeMaximum;
+
                 if (this.flightPhase === FLIGHT_PHASE.TAKEOFF || this.flightPhase === FLIGHT_PHASE.CLIMB) {
```

Comment:
yeah, i can do that. here's the new one:
```
    _calculateTargetedAltitudeVnav() {
        const hardRestrictedWaypointModel = this.fms.nextHardAltitudeRestrictedWaypoint;

        if (_isNil(hardRestrictedWaypointModel)) {
            return;
        }

        switch (this.flightPhase) {
            case FLIGHT_PHASE.TAKEOFF:
            case FLIGHT_PHASE.CLIMB:
                return this._calculateTargetedAltitudeVnavClimb(hardRestrictedWaypointModel);

            case FLIGHT_PHASE.CRUISE: {
                if (!this.isBeyondTopOfDescent()) {
                    return;
                }

                // Here we trigger the initial descent. Once vacating the filed cruise altitude, subsequent loops
                // will enter the below block because the flight phase will have become 'DESCENT'.
                return hardRestrictedWaypointModel.altitudeMaximum;
            }

            case FLIGHT_PHASE.DESCENT:
                return this._calculateTargetedAltitudeVnavDescent(hardRestrictedWaypointModel);

            default:
                return;
        }
    }
```

done